### ISSUE TITLE
Sending large payloads over WebSocket

### ISSUE THREAD

**Issue by erikvk** [Starcounter/Home#389]
@alemoi 

Question:

How would you send a large file stored on the local disk over a WebSocket connection? I want the entire file to be contained in a single message, but it may be split over multiple data frames if that is required.

**Comment by erikvk** [comment]
@alemoi 

+ A separate but related question:

If I assign a large `FileStream` to `Response.StreamedBody` in a HTTP response, but the client closes the TCP connection before the `Stream` is being read – will Starcounter still dispose the Stream (as it usually does)? The reason I ask is that I find residual temporary files on a server computer, that are removed when their corresponding `FileStream`s are disposed.

**Comment by alemoi** [comment]
I have looked on the existing C# APIs for WebSocket and they don't support streaming:
https://docs.microsoft.com/en-us/aspnet/core/fundamentals/websockets
https://docs.microsoft.com/en-us/dotnet/api/system.net.websockets.websocket?view=netframework-4.7.1

I would write a custom user protocol (unfortunately over several messages) that uses .NET async API and tasks to fetch data from files (with offsets, etc.) and send them gradually. This protocol should also handle WebSocket disconnects properly and close opened file streams, if any.

CC @miyconst maybe we have such an implementation already from some customer?

**Comment by miyconst** [comment]
We have an example for file upload, not for file download though. 

- [HandleFile registration](https://github.com/Starcounter/KitchenSink/blob/master/src/KitchenSink/Program.cs#L165).
- [FileUploadPage](https://github.com/Starcounter/KitchenSink/blob/1f4ff0f72fb0f101838a847f2f6d9df57ea4a5e9/src/KitchenSink/FileUploadPage.json.cs).
- [HandleFile class implementation](https://github.com/Starcounter/KitchenSink/blob/1f4ff0f72fb0f101838a847f2f6d9df57ea4a5e9/src/KitchenSink/HandleFile.cs).

It would be nice to have an example with file download as well though.

Keeping this issue around for time being.

**Comment by erikvk** [comment]
So is the fact that I cannot send multiple data frames within a single WebSocket message a restriction in Starcounter, or in the underlying technology?

**Comment by alemoi** [comment]
I was thinking about it and the reason why other WebSocket APIs like official .NET do not provide the streaming version: it requires a custom client-side protocol because one does not know automatically how to split stream into messages. 

> So is the fact that I cannot send multiple data frames within a single WebSocket message a restriction in Starcounter, or in the underlying technology?

I will check about how its done in other APIs about last frame.

**Comment by erikvk** [comment]
OK.

The way I see it, without deep understanding on how WebSockets are implemented in Starcounter, is that all I need is a way to define multiple data frames for a WebSocket message. I could imagine it being implemented like this:

```csharp
var fileStream = File.Create("some path");
var socket = new WebSocket(123);
var buffer = new byte[1024];
var message = new WebSocketMessage(socket);
using (fileStream)
using (var frameWriter = message.GetDataFrameWriter()) 
{
    while (fileStream.CanRead)
    {
        fileStream.Read(buffer, 0, 1024);
        frameWriter.WriteFrame(buffer, 0, 1024);
    }
}
```

**Comment by alemoi** [comment]
Yeah:

```
I would write a custom user protocol (unfortunately over several messages) that uses .NET async API and tasks to fetch data from files (with offsets, etc.) and send them gradually. This protocol should also handle WebSocket disconnects properly and close opened file streams, if any.
```
But then your client should be able to handle such split data. It will receive it in order and will need to append it simply. Probably such protocol would need to send data size first, and then needed amount of bytes, split into messages.

**Comment by miyconst** [comment]
Closing as won't fix in Starcounter 2.4 and not applicable in Starcounter 3.0.
